### Abgabe 6 Evgeniya Bujanowski

### 1. Schauen Sie sich das Jupyter Notebook aus der Veranstaltung „Daten managen I“ an und vollziehen die Schritte aus dem Live-Coding bis zur Zusammenführung der Umwelt- und Wetterdaten noch einmal nach. 

In [16]:
10 + 10

20

In [17]:
20  / 10

2.0

In [18]:
print("Hello World")

Hello World


In [19]:
pi = 3.14159

In [20]:
sentence = "Data literacy is useful."

In [21]:
sentence

'Data literacy is useful.'

In [22]:
len(sentence)

24

In [23]:
sentence.upper()

'DATA LITERACY IS USEFUL.'

In [24]:
sentence.count("a")

3

In [25]:
sentence.replace("a", "A")

'DAtA literAcy is useful.'

In [26]:
list_of_numbers = [5, 9, 12, 34]

In [27]:
list_of_numbers[0]

5

----------------------------------------------

### Reflektieren Sie Ihr Vorgehen und beantworten folgende Fragen: Was hat gut geklappt? Wenn Sie auf Fragen oder Probleme gestoßen sind, erläutern Sie diese kurz und beschreiben auch, welche Lösung Sie ggf. gefunden haben.

Alles hat soweit normal geklappt, ich habe keine Fragen.

----------------------------------------------

### 2. Laden Sie mittels Python die folgende Datei mit vom Umweltbundesamt erhobenen Messdaten für die Station Köln Rodenkirchen (DENW059) herunter: 

### https://raw.githubusercontent.com/konrad/DaLI_Basismodul_WiSe2021_2022/main/data/Rodenkirchen_processed.csv

### Lesen Sie die heruntergeladenen Daten, mithilfe der Programmbibliothek pandas, in einen DataFrame‚ "rodenkirchen" ein.

In [28]:
import pandas as pd

In [29]:
import urllib.request

In [30]:
rodenkirchen_url = "https://raw.githubusercontent.com/konrad/DaLI_Basismodul_WiSe2021_2022/main/data/Rodenkirchen_processed.csv"

In [31]:
rodenkirchen_file = "rodenkirchen.csv"

In [32]:
urllib.request.urlretrieve(rodenkirchen_url, rodenkirchen_file)

('rodenkirchen.csv', <http.client.HTTPMessage at 0x1f90d57df70>)

In [33]:
rodenkirchen = pd.read_csv(rodenkirchen_file)

In [34]:
rodenkirchen

Unnamed: 0 Stationscode               Datum  \
0              0      DENW059  '01.11.2020 01:00'   
1              1      DENW059  '01.11.2020 02:00'   
2              2      DENW059  '01.11.2020 03:00'   
3              3      DENW059  '01.11.2020 04:00'   
4              4      DENW059  '01.11.2020 05:00'   
...          ...          ...                 ...   
8732        8732      DENW059  '30.10.2021 22:00'   
8733        8733      DENW059  '30.10.2021 23:00'   
8734        8734      DENW059  '30.10.2021 24:00'   
8735        8735      DENW059  '31.10.2021 01:00'   
8736        8736      DENW059  '31.10.2021 02:00'   

     Feinstaub (PM₁₀) stündlich gleitendes Tagesmittel in µg/m³  \
0                                                    10           
1                                                    10           
2                                                    10           
3                                                    10           
4                                                    10           
...                                                 ...           
8732                                                 11           
8733                                                 11           
8734                                                 11           
8735                                                 10           
8736                                                 10           

     Ozon (O₃) Ein-Stunden-Mittelwert in µg/m³  \
0                                           30   
1                                           35   
2                                            -   
3                                           19   
4                                           10   
...                                        ...   
8732                                        31   
8733                                        29   
8734                                        16   
8735                                         7   
8736                                         4   

     Stickstoffdioxid (NO₂) Ein-Stunden-Mittelwert in µg/m³ Luftqualitätsindex  
0                                                    14               sehr gut  
1                                                     8               sehr gut  
2                                                    11               sehr gut  
3                                                     4               sehr gut  
4                                                     3               sehr gut  
...                                                 ...                    ...  
8732                                                 19               sehr gut  
8733                                                  -               sehr gut  
8734                                                 21                    gut  
8735                                                 17               sehr gut  
8736                                                 10               sehr gut  

[8737 rows x 7 columns]

------------------------

### Lassen Sie sich mit der method ‚describe‘ die Überblicksstatistiken zum DataFrame ‚rodenkirchen‘ anzeigen. 

read the documentation https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html

In [35]:
# Die Methode "describe" gibt ein Überblicksstatistiken für numerische Spalten zurück
rodenkirchen.describe()

Unnamed: 0
count  8737.000000
mean   4368.000000
std    2522.298984
min       0.000000
25%    2184.000000
50%    4368.000000
75%    6552.000000
max    8736.000000

Aus Interesse schau ich mir die ersten 100 Ergebnisse an, vllt ist da etwas Spannendes zu sehen

In [36]:
rodenkirchen.head(100)

Unnamed: 0 Stationscode               Datum  \
0            0      DENW059  '01.11.2020 01:00'   
1            1      DENW059  '01.11.2020 02:00'   
2            2      DENW059  '01.11.2020 03:00'   
3            3      DENW059  '01.11.2020 04:00'   
4            4      DENW059  '01.11.2020 05:00'   
..         ...          ...                 ...   
95          95      DENW059  '04.11.2020 24:00'   
96          96      DENW059  '05.11.2020 01:00'   
97          97      DENW059  '05.11.2020 02:00'   
98          98      DENW059  '05.11.2020 03:00'   
99          99      DENW059  '05.11.2020 04:00'   

   Feinstaub (PM₁₀) stündlich gleitendes Tagesmittel in µg/m³  \
0                                                  10           
1                                                  10           
2                                                  10           
3                                                  10           
4                                                  10           
..                                                ...           
95                                                 14           
96                                                 15           
97                                                 15           
98                                                 15           
99                                                 16           

   Ozon (O₃) Ein-Stunden-Mittelwert in µg/m³  \
0                                         30   
1                                         35   
2                                          -   
3                                         19   
4                                         10   
..                                       ...   
95                                         -   
96                                         -   
97                                         -   
98                                         -   
99                                         -   

   Stickstoffdioxid (NO₂) Ein-Stunden-Mittelwert in µg/m³ Luftqualitätsindex  
0                                                  14               sehr gut  
1                                                   8               sehr gut  
2                                                  11               sehr gut  
3                                                   4               sehr gut  
4                                                   3               sehr gut  
..                                                ...                    ...  
95                                                 27                    gut  
96                                                 25                    gut  
97                                                 20               sehr gut  
98                                                 18               sehr gut  
99                                                 16               sehr gut  

[100 rows x 7 columns]

Ich sehe z.B. jetzt, dass es Felder ohne Inhalt ("-") gibt. 
Ich kann noch einmal die Methode "describe" anwenden, allerdings dann die nicht existierenden Werte rausschmeißen.

In [37]:
rodenkirchen_decribe = pd.read_csv(rodenkirchen_file, na_values=["-"])

In [38]:
rodenkirchen_decribe.describe()

Unnamed: 0  \
count  8737.000000   
mean   4368.000000   
std    2522.298984   
min       0.000000   
25%    2184.000000   
50%    4368.000000   
75%    6552.000000   
max    8736.000000   

       Feinstaub (PM₁₀) stündlich gleitendes Tagesmittel in µg/m³  \
count                                        8691.000000            
mean                                           13.747900            
std                                             6.851536            
min                                             0.000000            
25%                                             9.000000            
50%                                            12.000000            
75%                                            17.000000            
max                                            65.000000            

       Ozon (O₃) Ein-Stunden-Mittelwert in µg/m³  \
count                                8129.000000   
mean                                   36.914208   
std                                    27.767770   
min                                     0.000000   
25%                                    13.000000   
50%                                    34.000000   
75%                                    56.000000   
max                                   161.000000   

       Stickstoffdioxid (NO₂) Ein-Stunden-Mittelwert in µg/m³  
count                                        8393.000000       
mean                                           21.746408       
std                                            13.391521       
min                                             0.600000       
25%                                            12.000000       
50%                                            19.000000       
75%                                            28.000000       
max                                           135.000000

### Wählen Sie eine der Variablen, geben Sie diese an und beschreiben Sie welche Informationen für diese Variable (Spalte) von ‚describe‘ geliefert werden.

Ich nehme die Spalte "Stickstoffdioxid (NO₂) Ein-Stunden-Mittelwert in µg/m³" (ab jetzt "Stickstoffdioxid" gennant). Dazu könnte ich Folgendes sagen:

count: 
Im Vergleich zu der Anzahl von Reihen im df rodenkirchen ist die count-Angabe kleiner - 8737.000000 zu 8393.000000. Ich gehe davon aus, dass der Grund ist, dass es in der Spalte "Stickstoffdioxid" na_values gab.

mean: 
Der Durchschnittswert (mean) liegt bei 21.746408. Trotzdem wäre es wichtig, auch den Medianwert zu ermitteln. Er ist, meine ich, verlässlicher als der Durchschnittswert, weil der Durchschnittswert durch ein paar Ausreißer geändert werden kann.

min: 
min liegt hier bei 0.600000. Ich verstehe es so, dass man immer Stickstoffdioxid in der Luft hat, nur manchmal sind die Mengen sehr klein.

---------------------------------------------------------

### 3. Diskutieren Sie, wann und warum es sinnvoll sein kann, für die Datenanalyse eine Programmiersprache wie Python oder ähnliche Lösungen zu nutzen. 

1) Es ist einfacher, mit Python automatisierte Prozesse zu entwickeln. Damit kann man sich wiederholende Aufgaben erledigen.
Mit Excel/ Excel VBA würde es unter Umständen auch gehen. Python ist eventuell mächtiges und schneller in dieser Hinsicht.

2) Python kann sehr große Datenmengen bearbeten, da kommt Excel je nach Aufgabe auf seine Grenzen.

3) Es gibt mehrere Bibliotheken für Visualisierungen in Python: plotly, seaborn, matpotlib etc. Damit lassen sich
überzeugene Visualisierungen machen. Je nach Bibliothek kann man auch animierte Visualisierunngen erzeugen. Wenn fertig, 
kann man eine Visualisierung als z.B. HTML oder als cvg (Scalable Vector Graphic) exportieren für weitere Schritte: 
zum Embedden auf einer Webseite oder zur weiteren Bearbeitung im Adobe Illustrator.

4) In Python kann man auch sentimant analysis machen, NLP Projekte durchführen. Diese Sachen sind interessant für Journalisten, aber auch für Marketing-Exprterten. 

5) Excel kann unter Umständen eine Zahl z.B. als Datum interprtieren. Das kann zu Fehlern führen. 

------------------------------------------------------

Random question: Wie kann die Luftqualität ausfallen?

In [39]:
rodenkirchen["Luftqualitätsindex"].unique()

array(['sehr gut', 'gut', 'mäßig', 'schlecht'], dtype=object)

Ok, die Luftqualitätsindex kann sehr gut, gut, mäßig und schlecht sein. 